File: python/questions/lengthOfLongestSubstring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        l = 0
        r = 0
        ext = set()
        while r < len(s):
            if s[r] in ext:
                ans = max(ans, r-l)
                while s[r] in ext:
                    ext.remove(s[l])
                    l += 1
            ext.add(s[r])
            r += 1
        ans = max(ans, r-l)
        return ans

File: rust/questions/length_of_longest_substring.rs
use std::collections::HashMap;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let s = s.as_bytes();
        let mut mp = HashMap::new();
        let mut l = 0;
        let mut r = 0;
        let mut ans = 0;
        while r < s.len() {
            let rc = s[r];
            if let Some(&pp) = mp.get(&rc) {
                ans = ans.max((r - l) as i32);
                l = l.max(pp + 1);
            }
            mp.insert(rc, r);
            r += 1;
        }
        ans = ans.max((r - l) as i32);
        ans
    }
}
